### PR TITLE
first version of elastic reindexer

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
First version of the elastic reindexer tool

-> has a `config.toml` where one can specify the source and the destination Elasticsearch instances data , alongside the indices to clone from source to destination

### CODE REVIEW

File: elasticreindexer/elastic/elasticClient.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+	if err != nil {
+		return err
+	}
+
+	if res.IsError() {
+		return fmt.Errorf("%s", res.String())
+	}
+
+	defer closeBody(res)
```
Comment (iulianpascalau) [inline elasticreindexer/elastic/elasticClient.go:98]: remove defer or move up

File: elasticreindexer/elastic/elasticClient.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+		numberOfHits := gjson.Get(string(scrollBodyBytes), "hits.hits.#")
+		if numberOfHits.Int() < 1 {
+			return nil
+		}
+		err := handlerFunc(scrollBodyBytes)
+		if err != nil {
+			return err
+		}
+
+		time.Sleep(1 * time.Second)
```
Comment (iulianpascalau) [inline elasticreindexer/elastic/elasticClient.go:199]: should extract this value in a const

File: elasticreindexer/elastic/elasticClient.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+		esc.client.ClearScroll.WithScrollID(scrollID),
+	)
+	if err != nil {
+		return err
+	}
+	if resp.IsError() && resp.StatusCode != http.StatusNotFound {
+		return fmt.Errorf("error response: %s", resp)
+	}
+
+	defer closeBody(resp)
```
Comment (iulianpascalau) [inline elasticreindexer/elastic/elasticClient.go:227]: remove defer or move up the call

File: elasticreindexer/process/serialization.go [hunk 4]
```diff
@@ -0,0 +1 @@
+package process
```
Comment (iulianpascalau) [inline elasticreindexer/process/serialization.go:1]: empty file

**Review by iulianpascalau — COMMENTED** [review]
Maybe in a future PR you can reference the indexer repo and remove the duplicated code

File: elasticreindexer/elastic/elasticClient.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+	scrollID string,
+	handlerFunc func(responseBytes []byte) error,
+) error {
+	if scrollID == "" {
+		return nil
+	}
+	defer func() {
+		err := esc.clearScroll(scrollID)
+		if err != nil {
+			log.Warn("cannot clear scroll ", err)
```
Comment (miiu96) [inline elasticreindexer/elastic/elasticClient.go:180]: log.Warn should have 3 arguments

File: elasticreindexer/elastic/elasticClient.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+		}
+	}()
+
+	for {
+		scrollBodyBytes, errScroll := esc.getScrollResponse(scrollID)
+		if errScroll != nil {
+			return errScroll
+		}
+
+		numberOfHits := gjson.Get(string(scrollBodyBytes), "hits.hits.#")
```
Comment (miiu96) [inline elasticreindexer/elastic/elasticClient.go:190]: gj

File: elasticreindexer/process/buffer.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
+package process
+
+import "bytes"
```
Comment (miiu96) [inline elasticreindexer/process/buffer.go:3]: you can reference this component from the **elastic-indexer-go** repository

File: elasticreindexer/process/queries.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+			"match_all": object{},
+		},
+	}
+
+	encoded, _ := encodeQuery(obj)
+
+	return &encoded
+}
+
+type generalElasticResponse struct {
```
Comment (miiu96) [inline elasticreindexer/process/queries.go:32]: `genericElasticResponse`

File: elasticreindexer/process/queries.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	Hits struct {
+		Hits []struct {
+			ID     string          `json:"_id"`
+			Source json.RawMessage `json:"_source"`
+		} `json:"hits"`
+	} `json:"hits"`
+}
+
+func extractSourceFromEsResponse(response generalElasticResponse) map[string]json.RawMessage {
+	recordsMap := make(map[string]json.RawMessage)
```
Comment (miiu96) [inline elasticreindexer/process/queries.go:42]: `recordsMap := make(map[string]json.RawMessage, len(hits))`

File: elasticreindexer/elastic/elasticClient.go (hunk 1 above)
Comment (bogdan-rosianu) [reply elasticreindexer/elastic/elasticClient.go:98]: done

File: elasticreindexer/elastic/elasticClient.go (hunk 2 above)
Comment (bogdan-rosianu) [reply elasticreindexer/elastic/elasticClient.go:199]: done

File: elasticreindexer/elastic/elasticClient.go (hunk 3 above)
Comment (bogdan-rosianu) [reply elasticreindexer/elastic/elasticClient.go:227]: done

File: elasticreindexer/process/serialization.go (hunk 4 above)
Comment (bogdan-rosianu) [reply elasticreindexer/process/serialization.go:1]: removed

File: elasticreindexer/cmd/elasticreindexer/config.toml [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+        username = ""
+        password = ""
+
+    [config.output]
+        url = "http://127.0.0.1:9200"
+        username = ""
+        password = ""
+
+    [config.indices]
+        indices = ["tps", "rating", "transactions", "blocks", "validators", "miniblocks", "rounds", "accounts", "accountshistory"]
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply elasticreindexer/cmd/elasticreindexer/config.toml:13]: added

File: elasticreindexer/elastic/elasticClient.go [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+		return 0, err
+	}
+
+	count := gjson.Get(string(respBytes), "count")
+	return count.Uint(), nil
+}
+
+// GetMapping will return the mapping of the specified index
+func (esc *esClient) GetMapping(index string) (*bytes.Buffer, error) {
+	res, err := esc.client.Indices.GetMapping(
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply elasticreindexer/elastic/elasticClient.go:66]: no need as the closing is done on `getBytesFromResponse`

File: elasticreindexer/elastic/elasticClient.go (hunk 5 above)
Comment (bogdan-rosianu) [reply elasticreindexer/elastic/elasticClient.go:180]: fixed

File: elasticreindexer/process/buffer.go (hunk 7 above)
Comment (bogdan-rosianu) [reply elasticreindexer/process/buffer.go:3]: will do this in a future PR

File: elasticreindexer/process/queries.go (hunk 8 above)
Comment (bogdan-rosianu) [reply elasticreindexer/process/queries.go:32]: i would keep the naming as it is

File: elasticreindexer/README.md [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# Elastic reindexer tool
```
Comment (iulianpascalau) [inline elasticreindexer/README.md:1]: :+1:

File: elasticreindexer/README.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+This tool is able to reindex all mappings and data from an instance of Elasticsearch to another.
+
+## How to use
+
+- update the Elasticsearch instance configuration for both source and destination inside `config.toml`;
+- update the indices list inside `config.toml`;
+- run `go build`;
+- run `./elasticreindexer` that will begin the process. Continuous prints should be observerd.
```
Comment (iulianpascalau) [inline elasticreindexer/README.md:10]: severd

File: elasticreindexer/process/queries.go (hunk 9 above)
Comment (bogdan-rosianu) [reply elasticreindexer/process/queries.go]: done

**Review by miiu96 — DISMISSED** [review]

**Review by iulianpascalau — DISMISSED** [review]

**Review by miiu96 — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**